File: client/javascript/src/prsim/sequence.ts
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

import {mpi} from "@/macro/mpi";
import {index} from "@/macro/idx";
import {spi} from "@/macro/spi";
import type {Context} from "@/prsim/context";
import {Status} from "@/cause";


@spi("mesh")
export abstract class Sequence {

    /**
     * 生成全网唯一序列号
     *
     * @param ctx ctx
     * @param kind 序列号类型，各个业务保持唯一.
     * @param length length
     * @return 唯一序列号
     */
    @mpi("mesh.sequence.next", String)
    next(ctx: Context, @index(0, 'kind') kind: string, @index(1, 'length') length: number): Promise<string> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * 获取序列号段.
     *
     * @param ctx ctx
     * @param kind 类型
     * @param size 号段大小
     * @param length length
     * @return 返回序列号列表
     */
    @mpi("mesh.sequence.section", [Array, String])
    section(ctx: Context, @index(0, 'kind') kind: string, @index(1, 'size') size: number, @index(2, 'length') length: number): Promise<string[]> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

}

File: client/java/src/main/java/com/be/mesh/client/prsim/Sequence.java
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package com.be.mesh.client.prsim;

import com.be.mesh.client.annotate.Index;
import com.be.mesh.client.annotate.MPI;
import com.be.mesh.client.annotate.SPI;

/**
 * @author coyzeng@gmail.com
 */
@SPI("mesh")
public interface Sequence {

    default String next(@Index(0) String kind) {
        return this.next(kind, 8);
    }

    /**
     * 生成全网唯一序列号
     *
     * @param kind   序列号类型，各个业务保持唯一.
     * @param length 长度
     * @return 唯一序列号
     */
    @MPI("mesh.sequence.next")
    String next(@Index(0) String kind, int length);

    /**
     * 获取序列号段.
     *
     * @param kind   类型
     * @param size   号段大小
     * @param length 长度
     * @return 返回序列号列表
     */
    @MPI("mesh.sequence.section")
    String[] section(@Index(0) String kind, @Index(1) int size, int length);

}
